File: Loenn/entities/SimpleText.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")
local drawableLine = require("structs.drawable_line")

local FemtoHelperSimpleText = {}

FemtoHelperSimpleText.name = "FemtoHelper/SimpleText"
FemtoHelperSimpleText.depth = -3000000
FemtoHelperSimpleText.texture = "loenn/FemtoHelper/simpletext"
FemtoHelperSimpleText.justification = {0.5, 0.5}
FemtoHelperSimpleText.nodeLimits = {0, 1}
FemtoHelperSimpleText.nodeVisibility = "always"
FemtoHelperSimpleText.nodeRenderType = "fan"

FemtoHelperSimpleText.fieldInformation = {
    layer = {
        options = {
            {"BelowBG", "BelowBG"},
            {"AboveBG", "AboveBG"},
            {"Gameplay", "Gameplay"},
            {"BelowFG", "BelowFG"},
            {"AboveFG", "AboveFG"},
            {"HUD", "HUD"},
            {"AdditiveHUD", "AdditiveHUD"}
        },
        editable = false
    },
    mainColor = {
        fieldType = "color",
        useAlpha = true
    },
    outlineColor = {
        fieldType = "color",
        useAlpha = true
    }
}

FemtoHelperSimpleText.placements = {
    name = "default",
    data = {
        depth = -10000,
        fontDataPath = "Graphics/FemtoHelper/PlutoniumText/Example.xml",
        justifyX = 0.5,
        justifyY = 0.5,
        dialogID = "FemtoHelper_PlutoniumText_Example",
        --hud = true,
        truncateSliderValues = false,
        mainColor = "FFFFFFFF",
        outlineColor = "000000FF",
        parallax = 1,
        scale = 1,
        shadow = false,
        extraSpacing = 0,
        visibilityFlag = "",
        decimals = -1,
        layer = "AboveFG",
        hudZoomSupport = true,
    }
}

function FemtoHelperSimpleText.nodeSprite(room, entity, node) 
    local spr = {}
    local sprite = drawableSprite.fromTexture("loenn/FemtoHelper/simpletextnode", entity)
    sprite:addPosition(node.x - entity.x, node.y - entity.y);
    
    table.insert(spr, sprite)
    table.insert(spr, drawableLine.fromPoints({entity.x, entity.y, node.x, node.y}, {1, 0, 0.5, 0.3}, 1))

    return spr
end

function FemtoHelperSimpleText.sprite(room, entity, node, nodeIndex)
    local main = drawableSprite.fromTexture("loenn/FemtoHelper/simpletext", entity)
    local outline = drawableSprite.fromTexture("loenn/FemtoHelper/simpletextOutline", entity)

    main:setColor(entity.mainColor)
    outline:setColor(entity.outlineColor)

    return {
        main,
        outline
    }
end

function FemtoHelperSimpleText.selection(room, entity)
    local nodelist = {}
    for _, node in ipairs(entity.nodes) do

        table.insert(nodelist, utils.rectangle(node.x - 4, node.y - 4, 8, 8))

    end
    return utils.rectangle(entity.x - 8, entity.y - 4, 16, 8), nodelist
end

return FemtoHelperSimpleText

File: Source/PlutoniumText/SimpleText.cs
using Celeste.Mod.FemtoHelper.PlutoniumText;
using Celeste.Mod.FemtoHelper.Utils;
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Celeste.Mod.FemtoHelper.Entities;

[CustomEntity("FemtoHelper/SimpleText")]
public partial class SimpleText : Entity
{
    public readonly List<PlutoniumTextNodes.Node> Nodes;
    public Color Color1;
    public Color Color2;
    public readonly bool Shadow;
    public readonly int Spacing;
    public readonly PlutoniumTextComponent Text;
    public readonly float Parallax;
    public bool Hud => Text?.Layer == TextLayer.HUD || Text?.Layer == TextLayer.AdditiveHUD;

    public readonly bool TruncateSliders;
    public readonly int Decimals;
    public readonly float Scale;
    public readonly string VisibilityFlag;

    public readonly Vector2 RenderOffset;

    private readonly bool legacy;

    public readonly Vector2 Justify;

    public readonly bool HudCanZoom;
    public float ZoomFactor => !Hud ? 1 : (HudCanZoom ? 1 : SceneAs<Level>().Zoom);
    public SimpleText(EntityData data, Vector2 offset) : base(data.Position + offset)
    {
        TruncateSliders = data.Bool("truncateSliderValues", false);
        Decimals = data.Int("decimals", -1);

        Nodes = [];
        if (!Dialog.Has(data.Attr("dialogID", "FemtoHelper_PlutoniumText_Example")))
        {
            Nodes.Add(new PlutoniumTextNodes.Text(Dialog.Get(data.Attr("dialogID", "FemtoHelper_PlutoniumText_Example"))));
        }
        else
        {
            Nodes = PlutoniumTextNodes.Parse(data.Attr("dialogID", "FemtoHelper_PlutoniumText_Example"), TruncateSliders, Decimals);
        }

        Color1 = Calc.HexToColorWithAlpha(data.Attr("mainColor", "ffffffff"));
        Color2 = Calc.HexToColorWithAlpha(data.Attr("outlineColor", "000000ff"));
        Depth = data.Int("depth", -100);
        Shadow = data.Bool("shadow", false);
        string list = data.Attr("charList", "");
        legacy = !string.IsNullOrWhiteSpace(list);
        Vector2 size = new(data.Int("fontWidth", 8), data.Int("fontHeight", 12));
        //Hud = data.Bool("hud", false);

        Justify = data.Vector2("justifyX", "justifyY", new Vector2(0.5f, 0.5f));

        string fontFilePath = data.Attr("fontDataPath", data.Attr("fontPath", "objects/FemtoHelper/PlutoniumText/example"));

        Spacing = legacy ? data.Int("spacing", 7) - (int)size.X : 0;
        Spacing += data.Int("extraSpacing", 0);

        TextLayer layer = data.Enum("layer", data.Bool("hud", false) ? TextLayer.HUD : TextLayer.Gameplay);
        Add(Text = new PlutoniumTextComponent(fontFilePath, layer, null, RenderCallback, new(), list, size));
        Parallax = data.Float("parallax", 1);

        HudCanZoom = data.Bool("hudZoomSupport", false);

        Scale = data.Float("scale", 1);
        VisibilityFlag = data.Attr("visibilityFlag", "");

        RenderOffset = (data.FirstNodeNullable(offset) ?? Position) - Position;

        if (Hud) Tag |= TagsExt.SubHUD;
    }

    //public override void Update()
    //{
    //    base.Update();
    //    Log(Text.Layer.ToString());
    //}

    public override void Render()
    {
        base.Render();
        if (Text.Layer == TextLayer.Gameplay) RenderCallback(SceneAs<Level>());
    }

    public void RenderCallback(Level level)
    {
        if (!level.FancyCheckFlag(VisibilityFlag)) return;

        bool flip = (SaveData.Instance?.Assists.MirrorMode ?? false);

        bool hudFlip = flip && Hud;

        float ww = 160f / level.Zoom;
        float hh = 90f / level.Zoom;

        Vector2 camPos = level.Camera.Position;
        Vector2 camCenter = camPos + new Vector2(ww, hh);
        Vector2 position2 = (((Position - camCenter) * Parallax / ZoomFactor) * new Vector2(hudFlip ? -1 : 1, 1)) + camCenter;

        string str2 = PlutoniumTextNodes.ConstructString(Nodes, SceneAs<Level>());

        Rectangle visRect = Text.GetVisibilityRectangle(position2 + RenderOffset, str2, Spacing, Scale, Justify, ZoomFactor);

        if (!visRect.IsVisible()) return;

        float scale2 = Scale;

        if (Hud)
        {
            position2 *= 6f;
            scale2 *= 6f / ZoomFactor;
        }

        Text.Print(position2 + (RenderOffset * (Hud ? 6f : 1f)), str2, Shadow, Spacing, Color1, Color2, Justify, scale2, 0, flip && (Text.Layer != TextLayer.HUD));
    }

    public override void DebugRender(Camera camera)
    {
        base.DebugRender(camera);
        bool flip = (SaveData.Instance?.Assists.MirrorMode ?? false);

        bool hudFlip = flip && Hud;

        Level level = SceneAs<Level>();

        float ww = 160f / level.Zoom;
        float hh = 90f / level.Zoom;

        Vector2 camPos = camera.Position;
        Vector2 camCenter = camPos + new Vector2(ww, hh);
        Vector2 position2 = (((Position - camCenter) * Parallax / ZoomFactor) * new Vector2(hudFlip ? -1 : 1, 1)) + camCenter;

        Vector2 p2f = position2.Floor();

        Draw.HollowRect(position2.X - 2f, position2.Y - 2f, 4f, 4f, Color.BlueViolet);

        Draw.HollowRect(Text.GetVisibilityRectangle(p2f + RenderOffset, PlutoniumTextNodes.ConstructString(Nodes, SceneAs<Level>()), Spacing, Scale, Justify, ZoomFactor), Color.MediumOrchid * 0.5f);
    }

    [GeneratedRegex(@"(\{|\})")]
    public static partial Regex MyRegex();
}
